tfjs-backend-webgpu/src/kernels/concat_webgpu.ts

```diff
@@ -42,32 +43,28 @@ export class ConcatProgram implements WebGPUProgram {
         this.dispatchLayout, this.outputShape, this.workGroupSize,
         [this.workPerThread, 1, 1]);
 
-    this.shapes = shapes;
-    // shapes is used by const snippets.
-    this.shaderKey = `concat${shapes}`;
+    this.offsetLength = shapes.length - 1;
+    for (let i = 0; i < this.offsetLength; i++) {
+      this.uniforms += `offset${i} : i32;`;
+    }
+    this.shaderKey = 'concat';
   }
 
   getUserCode(): string {
-    const offsets: number[] = new Array(this.shapes.length - 1);
     const snippets: string[] = [];
-    if (offsets.length > 0) {
-      offsets[0] = this.shapes[0][1];
-      for (let i = 1; i < offsets.length; i++) {
-        offsets[i] = offsets[i - 1] + this.shapes[i][1];
-      }
-
-      snippets.push(`if (yC < ${
-          offsets[0]}){ setOutput(coords.x, coords.y, getT0(yR, yC)); }`);
-      for (let i = 1; i < offsets.length; i++) {
-        const shift = offsets[i - 1];
+    if (this.offsetLength > 0) {
+      snippets.push(
+          `if (yC < uniforms.offset0){ setOutput(coords.x, coords.y, getT0(yR, yC)); }`);
+      for (let i = 1; i < this.offsetLength; i++) {
         snippets.push(
-            `elseif (yC < ${offsets[i]}){ ` +
-            `setOutput(coords.x, coords.y, getT${i}(yR, yC - ${shift})); }`);
+            `elseif (yC < uniforms.offset${[i]}){ ` +
+            `setOutput(coords.x, coords.y, getT${i}(yR, yC - uniforms.offset${
+                i - 1})); }`);
       }
-      const lastIndex = offsets.length;
-      const lastShift = offsets[offsets.length - 1];
+      const lastIndex = this.offsetLength;
+      const lastShift = this.offsetLength - 1;
```

Comment:
nit: lastShiftIndex?

Reply:
I don't know the difference between the two, so I prefer to just keep it.